### PR TITLE
chore(atomix): remove heartbeat thread

### LINKED ISSUE #4200: Remove raft heartbeat thread in followers
**Description**

The main reason for adding a thread for handling heartbeat was that when we did not provision enough resources in our cluster, the append to log was sometimes taking too long that the leader cannot send heartbeats in time. Followers then hit the election timeout and start a new election even though the leader is reachable.

Having another thread to handle heartbeats has introduced lots of concurrency bugs (#4137 , https://github.com/zeebe-io/atomix/issues/139). It would be safe to remove the heartbeat thread in the followers because it’s only a problem if the leader cannot send heartbeats in time. If follower append takes too long, it can still cancel the heartbeat timer which will cancel the scheduled handler to start election.

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

- Removes heartbeat threadcontext
- Removes heartbeat request and only use append requests as heartbeats

## Related issues

closes #4200 #4201 #4137 

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java [hunk 1]
```diff
@@ -88,14 +86,14 @@ public FollowerRole(final RaftContext context) {
   @Override
   public CompletableFuture<InstallResponse> onInstall(final InstallRequest request) {
     final CompletableFuture<InstallResponse> future = super.onInstall(request);
-    resetHeartbeatTimeoutFromDifferentThread();
+    if (isRequestFromCurrentLeader(request.term(), request.leader())) {
```
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java:89]: :+1:

File: atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java [hunk 2]
```diff
@@ -188,61 +186,50 @@ private void sendPollRequests() {
...
   }
 
   @Override
   public CompletableFuture<AppendResponse> onAppend(final AppendRequest request) {
-    // Reset the heartbeat timeout.
-    resetHeartbeatTimeoutFromDifferentThread();
-
-    return super.onAppend(request);
+    final CompletableFuture<AppendResponse> future = super.onAppend(request);
+    if (isRequestFromCurrentLeader(request.term(), request.leader())) {
```
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java:198]: Should we reset it after the append is complete? Shouldn't we acknowledge it immediately?

**Review by npepinpe — COMMENTED** [review]
Looks good in general, just some questions.

I know we talked about this, just want to make sure again, please bear with me :sweat_smile: - is that we cancel the heartbeat after we've processed the request. This is safe again because...? Because even if the heartbeat timer expires and the task is scheduled, if we cancel it before the executor continues, it is correctly cancelled?

Also, that we reuse the same timer for the poll should now be pretty safe as everything is done on the same thread, correct?

Finally, do we have tests for this? I couldn't find anything in `RaftTest` but maybe I'm not looking in the right place.

**Comment by deepthidevaki** [comment]
> Looks good in general, just some questions.
> 
> I know we talked about this, just want to make sure again, please bear with me sweat_smile - is that we cancel the heartbeat after we've processed the request. This is safe again because...? Because even if the heartbeat timer expires and the task is scheduled, if we cancel it before the executor continues, it is correctly cancelled?

Yes. When the task is executed, it check for its current state which should be cancelled if we cancelled it in the append.

> 
> Also, that we reuse the same timer for the poll should now be pretty safe as everything is done on the same thread, correct?
> 

Yes. We always cancel the current timer before we reset it. So I guess it is safe and we never have dangling timers.

> Finally, do we have tests for this? I couldn't find anything in `RaftTest` but maybe I'm not looking in the right place.

I did not add any test specific to this. Didn't really think about it. Isn't the existing leader failover tests enough for this? What specific scenarios should we test?

**Comment by deepthidevaki** [comment]
> > Looks good in general, just some questions.
> > I know we talked about this, just want to make sure again, please bear with me sweat_smile - is that we cancel the heartbeat after we've processed the request. This is safe again because...? Because even if the heartbeat timer expires and the task is scheduled, if we cancel it before the executor continues, it is correctly cancelled?
> 
> Yes. When the task is executed, it check for its current state which will be 'cancelled' if we cancelled it in the append.

Also it is safer to reset the time after append completes because if we reset the timer first and the append takes longer than the timer it might trigger the timer immediately after the append.

**Comment by npepinpe** [comment]
I actually found one test in `RaftTest` for heartbeats, after giving it a second look: `RaftTest#testThreeNodeManyEventsDoNotMissHeartbeats`, which counts that we never miss heartbeats using metrics. Do we have failover tests for heartbeat misses? I only found failover tests when we explicitly close the leader but not when the leader is still alive. Would you say that is equivalent? On graceful shutdown we do send a cluster membership event so I'm not sure that would test the same thing. It's also possible I missed some tests and we do have something covering the case where a leader is too slow, or has crashed.

**Comment by deepthidevaki** [comment]
Actually all failovers in RaftTest should be due to heartbeat misses because it uses a mock ClusterMembershipService. If we need a test to explicitly test for hearbeat misses, we would need to simulate network partition I guess.

**Review by npepinpe — APPROVED** [review]

**Pull request merged**